lexer/stateful/pattern.go

```diff
@@ -0,0 +1,208 @@
+		res, _ := getPotentialStarts(r.Sub[0])
+		return res, mayMatchZero
+
+	case syntax.OpRepeat:
+		res, nb := getPotentialStarts(r.Sub[0])
+		if r.Min == 0 {
+			return res, mayMatchZero
+		} else if r.Min == 1 && r.Max == 1 {
+			return res, nb
+		}
+		return res, mayMatchOneOrMore
+
+	case syntax.OpCapture:
+		return getPotentialStarts(r.Sub[0])
+
+	case syntax.OpPlus:
+		var nodes, nb = getPotentialStarts(r.Sub[0])
+		return nodes, min(nb, mayMatchOneOrMore) // (^\b+) is a potential zero length match
+
+	case syntax.OpLiteral:
+		// Literal can also visibly be a sequence of characters, hence the check on the length
+		// of the Runes.
+		res := []*syntax.Regexp{r}
+		if len(r.Rune) > 1 {
+			return res, mayMatchOneOrMore
+		}
+		return res, matchesExactlyOne
+
+	case syntax.OpCharClass, syntax.OpAnyChar, syntax.OpAnyCharNotNL:
+		// These are one character matches.
+		return []*syntax.Regexp{r}, matchesExactlyOne
+	}
+
+	panic("should not get here")
+}
+
+// appendRange returns the result of appending the range lo-hi to the class r.
+// function stolen from syntax/parse.go
+func appendRange(r []int, lo, hi int) []int {
+	// Expand last range or next to last range if it overlaps or abuts.
+	// Checking two ranges helps when appending case-folded
+	// alphabets, so that one range can be expanding A-Z and the
+	// other expanding a-z.
+	n := len(r)
+	for i := 2; i <= 4; i += 2 { // twice, using i=2, i=4
+		if n >= i {
+			rlo, rhi := r[n-i], r[n-i+1]
+			if lo <= rhi+1 && rlo <= hi+1 {
+				if lo < rlo {
+					r[n-i] = lo
+				}
+				if hi > rhi {
+					r[n-i+1] = hi
+				}
+				return r
+			}
+		}
+	}
+
+	return append(r, lo, hi)
+}
+
+func computeNumberOfRunes(arg []int) int {
+	var acc = 0
+	for i, l := 0, len(arg); i < l; i += 2 {
+		acc += arg[i+1] - arg[i]
+	}
+	return acc
+}
+
+type computedRuneRange struct {
+	pattern string
+	runes   []int
+	size    int
+	nbmatch int
+}
+
+// computeRuneRanges gives the rune ranges that are potential starts for the regexp
+// it returns the ranges in a single slice as [start,stop,start,stop,...]. The ranges
+// are inclusive.
+// The second return value is the number of runes included.
+// FIXME ; the ranges should be simplified when added
+func computeRuneRanges(pattern string) (*computedRuneRange, error) {
+	var (
+		syn   *syntax.Regexp
+		res   []int
+		count = 0
+		err   error
+	)
+
+	if syn, err = syntax.Parse(pattern, syntax.Perl); err != nil {
+		return nil, fmt.Errorf("could not compute rune ranges for pattern /%s/: %w", pattern, err)
+	}
+	// this step is not required and may actually take up time for nothing, as the pattern seems to be
+	// somewhat simplified by default
+	// syn = syn.Simplify()
+
+	// should also check for match length
+	var startNodes, nbmatch = getPotentialStarts(syn)
+
+	if nbmatch == mayMatchZero || startNodes == nil {
+		// If there is at least one non-definite match, then we have an error on our hands !
+		// FIXME this has to be handled
```

Comment:
What has to be handled?

Reply:
This statement is to raise an error if a pattern may create a zero-length match.

Reply:
Not relevant any more ?